### ISSUE TITLE
4.12 - Windows: Crash in caml_memprof_handle_postponed_exn

### ISSUE THREAD

**Issue by bryphe** [ocaml/ocaml#10422]
Hi!

I just recently upgraded my project to OCaml 4.12 from 4.11, and received a user report of a crash on Windows after this upgrade (https://github.com/onivim/oni2/issues/3579):

Hooking up WinDbg and diving in - I saw that the crash is occuring in [`caml_memprof_handle_postponed_exn`](https://github.com/ocaml/ocaml/blob/cce52acc7c7903e92078e9fe40745e11a1b944f0/runtime/memprof.c#L572). 

Specifically, in this code: 
```
/* In case of a thread context switch during a callback, this can be
   called in a reetrant way. */
value caml_memprof_handle_postponed_exn(void)
{
  value res = Val_unit;
  uintnat i;
  if (local->suspended) return Val_unit; 
  if (callback_idx >= entries_global.len && local->entries.len == 0)
    return Val_unit;
```

`local` is NULL, and results in an access violation.

One curious aspect is that this is occurring on the main thread, where `local` should always point to the main-thread context, if my understanding of [this code](https://github.com/ocaml/ocaml/blob/cce52acc7c7903e92078e9fe40745e11a1b944f0/runtime/memprof.c#L167) is correct.

My suspicion is that, somehow, the `local` is being inadvertently cleared out - potentially by the [`caml_memprof_leave_thread` function]:(https://github.com/ocaml/ocaml/blob/cce52acc7c7903e92078e9fe40745e11a1b944f0/runtime/memprof.c#L1126)
```
CAMLexport void caml_memprof_leave_thread(void)
{
  local = NULL;
}
```

I'm not sure yet, though, if this is a bug in my code or a dependent library - like, an improper use of the `caml_acquire_runtime_system`/`caml_release_runtime_system` primitives. It seems like an allocation happening after releasing the runtime could potentially hit this, too. At first, I suspected maybe it was a Thread not registering with the runtime system, but once I realized the exception was happening on the main thread, I ruled this out.

It's also strange that I'm not seeing this on OSX or Linux - just Windows.

<details>
<summary>WinDBG info</summary>

From WinDBG, the symbols are a bit mangled, but this is what I get:
```
(65d4.64ec): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
Oni2_editor!tree_sitter_c+0x2e2ac4:
00000000`00d84b74 8b10            mov     edx,dword ptr [rax] ds:00000000`00000000=????????
0:000> k
 # Child-SP          RetAddr           Call Site
00 00000000`05ecebf0 00000000`00d65615 Oni2_editor!tree_sitter_c+0x2e2ac4
01 00000000`05ecec40 00000000`00d6a867 Oni2_editor!tree_sitter_c+0x2c3565
02 00000000`05ecec70 00000000`00d86541 Oni2_editor!tree_sitter_c+0x2c87b7
03 00000000`05ecece0 00000000`00000001 Oni2_editor!tree_sitter_c+0x2e4491
04 00000000`05ecece8 00000000`28815058 0x1
05 00000000`05ececf0 00000000`0dd66030 0x28815058
06 00000000`05ececf8 00000000`00000001 0xdd66030
07 00000000`05eced00 40906800`00000000 0x1
08 00000000`05eced08 40806000`00000000 0x40906800`00000000
09 00000000`05eced10 00000000`00000000 0x40806000`00000000
```

Cross-referencing the addresses with `uf` and `x86_64-w64-mingw32-nm`:
```
0:000> uf Oni2_editor!tree_sitter_c+0x2e2ac4
Oni2_editor!tree_sitter_c+0x2e2ab0:
00000000`00d84b60 55              push    rbp
00000000`00d84b61 57              push    rdi
00000000`00d84b62 56              push    rsi
00000000`00d84b63 53              push    rbx
00000000`00d84b64 4883ec28        sub     rsp,28h
00000000`00d84b68 488b0579dc8800  mov     rax,qword ptr [Oni2_editor!tree_sitter_c+0xb70738 (00000000`016127e8)]
00000000`00d84b6f bf01000000      mov     edi,1
00000000`00d84b74 8b10            mov     edx,dword ptr [rax]    <-- CRASH HERE
00000000`00d84b76 85d2            test    edx,edx
00000000`00d84b78 0f8559010000    jne     Oni2_editor!tree_sitter_c+0x2e2c27 (00000000`00d84cd7)  Branch

Oni2_editor!tree_sitter_c+0x2e2ace:
00000000`00d84b7e 488b35bb5fef00  mov     rsi,qword ptr [Oni2_editor!tree_sitter_c+0x11d8a90 (00000000`01c7ab40)]
00000000`00d84b85 483935ccdc8800  cmp     qword ptr [Oni2_editor!tree_sitter_c+0xb707a8 (00000000`01612858)],rsi
00000000`00d84b8c 770b            ja      Oni2_editor!tree_sitter_c+0x2e2ae9 (00000000`00d84b99)  Branch
```
```
PS C:\oni2\_esy\default\store\i\oni2-093a99ea\bin> x86_64-w64-mingw32-nm.exe Oni2_editor.exe | sls  00d84b60

0000000000d84b60 T caml_memprof_handle_postponed_exn
```
</details>

Seems related to the change in #9674 

Still digging to see if I can narrow this down on my end, but figured I'd share the info I have so far -

**Comment by bryphe** [comment]
Debugging further - it looks like this consistently occuring in `SDL_PollEvent`: https://github.com/revery-ui/revery/blob/4c2bdac424901f7b112b73175366e46135d26cdc/packages/reason-sdl2/src/sdl2_wrapper.cpp#L1029

It looks like the culprit is that, there is a callback that can potentially be dispatched during that call (w/o) the runtime: https://github.com/revery-ui/revery/blob/4c2bdac424901f7b112b73175366e46135d26cdc/packages/reason-sdl2/src/sdl2_wrapper.cpp#L158

In other words, the `SDL_PollEvent` triggers a C callback which we registered - but the problem is the runtime has been released, and any allocation would blow up because `local` is set to NULL.
